### PR TITLE
sql, server: setup periodic async job that performs temp table cleanup

### PR DESCRIPTION
**Pull request by arulajmani**
Previously, every session deleted temporary tables created by it when
it exited gracefully. However, if the session did not exit gracefully,
temporary table metadata + descriptors would be left hanging and
innacessible forever. This PR addresses this.

After this patch, when the server starts for the first time, it
configures a background deletion job for temporary tables. This job is
run asynchronously every 30 minutes. It ensures that every temporary
schema entry in system.namespac belongs to an active session. If that
is not the case, the temporary schema, and everything under it, must be
deleted.

Release note (sql): Temporary tables created by sessions that exit
ungracefully, and therefore are unable to perform cleanup, are
periodically cleaned up now.

### CODE REVIEW

File: pkg/sql/temporary_tables_deletion.go [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+func (concurrentDeleteTempTablesError) Error() string {
+	return "another delete temp tables job is already running"
+}
+
+// ConcurrentDeleteTempTablesError is reported when two delete temp tables jobs
+// are issued concurrently. This is a sentinel error.
+var ConcurrentDeleteTempTablesError error = concurrentDeleteTempTablesError{}
+
+func checkRunningTempTablesJob(ctx context.Context, job *jobs.Job, p *planner) error {
```
Comment (solongordon) [inline pkg/sql/temporary_tables_deletion.go]: This function is pretty much copy-pasted from `checkRunningJobs`. Could you refactor to share more code?

File: pkg/sql/temporary_tables_deletion.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+		Progress:    jobspb.DeleteTempTablesProgress{},
+	}
+
+	job, errCh, err := p.ExecCfg().JobRegistry.StartJob(ctx, resultsCh, record)
+	if err != nil {
+		return err
+	}
+
+	if err = <-errCh; err != nil {
+		if errors.Is(err, ConcurrentDeleteTempTablesError) {
```
Comment (solongordon) [inline pkg/sql/temporary_tables_deletion.go]: This is a compile error for me. I think you want to import `github.com/cockroachdb/errors`?

File: pkg/sql/temporary_schema.go [hunk 3]
```diff
@@ -47,8 +50,25 @@ func (p *planner) createSchemaWithID(
 	return p.txn.Run(ctx, b)
 }
 
+func temporarySchemaSessionID(scName string) (bool, ClusterWideID, error) {
```
Comment (solongordon) [inline pkg/sql/temporary_schema.go]: Could use a comment on this function and what it returns.

**Review by solongordon — CHANGES_REQUESTED** [review]
I did an initial scan. Any thoughts on how to test this?

By the way, is the way this works that all the nodes try to run this job, and it fails if the job is already in progress on another node? So it might run more often than every 30 minutes? I don't think that's necessarily a problem, just want to understand what's happening.

**Review by arulajmani — COMMENTED** [review]
Testing: 
I haven't done this yet, but it might help if we could add a knob that decides wether to perform cleanup at session exit or not. We could also make the time between session deletion jobs configurable, which would allow us to test the whole process end to end. 

Job Frequency:
I think it'll fail if it sees the job is in progress on another node, but depending on how these times line up, we could still be running the job more often than 30 minutes (changed this to 2 hours now, btw). I think there is opportunity to build something fancier around here, not sure how important that would be. 

PR Thoughts:
I've rebased + refactored the code to use the new `cleanupSessionTempObjects` implementation from the previous PR. The new implementation uses an executor to drop the tables. Earlier, the jobs framework didn't have the ability to create a `SessionBoundInternalExecutor`.  We couldn't use the `InternalExecutor` interface, because we need to set session data with the appropriate `SearchPath` to drop temporary objects. So I had to add a factory to the job registry to allow creation for these `SessionBoundInternalExecutors`.  
Another note on this factory -- I chose to add this field after the jobRegistry was created, and not during construction. This is because there seems to be an enforced initialization ordering in `server/server.go`. Moving things around was causing seg faults, though I didn't spend too much time trying to understand why/trying all possible combinations. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42919#-:-LwZj0n3-P1-yS51xaxz:b-9oq42o)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @solongordon)

___
*[pkg/sql/descriptor.go, line 318 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42919#-LvC1QgYYEKzv1m3czVM-r3-318:-LwZhA_D1z1QDh5eT3wr:bvyytxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/1953211aebe3328201a586d38765e894f86ed875/pkg/sql/descriptor.go#L318)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Do you actually need all the columns? Seems like you're only reading one of them.
</blockquote></details>

Yeah, I just need the "name" column. Is there a way to just read a single column? 
___
*[pkg/sql/descriptor.go, line 350 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42919#-LvC1QgYYEKzv1m3czVM-r3-350:-LwZhJv-5cgV6h3xbccw:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1953211aebe3328201a586d38765e894f86ed875/pkg/sql/descriptor.go#L350)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Please make the 2 a constant rather than a magic number.
</blockquote></details>

Done. 
___
*[pkg/sql/temporary_schema.go, line 53 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42919#-LvC1Qgbdahm0eaanf7m-r3-53:-LwPypDL8r_9zlqPX87Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1953211aebe3328201a586d38765e894f86ed875/pkg/sql/temporary_schema.go#L53)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Could use a comment on this function and what it returns.
</blockquote></details>

Done.
___
*[pkg/sql/temporary_tables_deletion.go, line 108 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42919#-LvC1QgkHyY0AIrv8BHw-r3-108:-LwZcALj5fRThzyuSKJv:b5ccjq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1953211aebe3328201a586d38765e894f86ed875/pkg/sql/temporary_tables_deletion.go#L108)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

This function is pretty much copy-pasted from `checkRunningJobs`. Could you refactor to share more code?
</blockquote></details>

Had this as a mental note, totally forgot to do it before I pushed! Done now! 
___
*[pkg/sql/temporary_tables_deletion.go, line 165 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42919#-LvC1QgkHyY0AIrv8BHw-r3-165:-LwZat6NDl2FZb4JyQIn:b-cqvar3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1953211aebe3328201a586d38765e894f86ed875/pkg/sql/temporary_tables_deletion.go#L165)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

This is a compile error for me. I think you want to import `github.com/cockroachdb/errors`?
</blockquote></details>

Weird, for me importing "errors" seems to be okay and this compiles.

**Comment by otan** [comment]
superseded by #45185

**Pull request closed without merging**